pkg/gen/nodejs-templates/helm/v2/helm.ts

```diff
@@ -233,77 +233,27 @@ export class Chart extends yaml.CollectionComponentResource {
         // overrides its predecessory, rather than throwing an exception, and `schema:
         // jsyaml.CORE_SCHEMA` to avoid using additional YAML parsing rules not supported by the
         // YAML parser used by Kubernetes.
-        const objs = yamlStream.split(/^---/m)
-            .map(yaml => jsyaml.safeLoad(yaml, {json: true, schema: jsyaml.CORE_SCHEMA}))
-            .filter(a => a != null && "kind" in a)
-            .sort(helmSort);
-        return yaml.parse(
-            {
-                resourcePrefix: resourcePrefix,
-                yaml: objs.map(o => jsyaml.safeDump(o)),
-                transformations: transformations || [],
-            },
-            { parent: this, dependsOn: dependsOn }
-        );
-    }
-}
-
-// helmSort is a JavaScript implementation of the Helm Kind sorter[1]. It provides a
-// best-effort topology of Kubernetes kinds, which in most cases should ensure that resources
-// that must be created first, are.
-//
-// [1]: https://github.com/helm/helm/blob/094b97ab5d7e2f6eda6d0ab0f2ede9cf578c003c/pkg/tiller/kind_sorter.go
-/** @ignore */ export function helmSort(a: { kind: string }, b: { kind: string }): number {
-    const installOrder = [
-        "Namespace",
-        "ResourceQuota",
-        "LimitRange",
-        "PodSecurityPolicy",
-        "Secret",
-        "ConfigMap",
-        "StorageClass",
-        "PersistentVolume",
-        "PersistentVolumeClaim",
-        "ServiceAccount",
-        "CustomResourceDefinition",
-        "ClusterRole",
-        "ClusterRoleBinding",
-        "Role",
-        "RoleBinding",
-        "Service",
-        "DaemonSet",
-        "Pod",
-        "ReplicationController",
-        "ReplicaSet",
-        "Deployment",
-        "StatefulSet",
-        "Job",
-        "CronJob",
-        "Ingress",
-        "APIService"
-    ];
-
-    const ordering: { [key: string]: number } = {};
-    installOrder.forEach((_, i) => {
-        ordering[installOrder[i]] = i;
-    });
-
-    const aKind = a["kind"];
-    const bKind = b["kind"];
-
-    if (!(aKind in ordering) && !(bKind in ordering)) {
-        return aKind.localeCompare(bKind);
-    }
-
-    if (!(aKind in ordering)) {
-        return 1;
-    }
-
-    if (!(bKind in ordering)) {
-        return -1;
+        const promise = pulumi.output(pulumi.runtime.invoke(
+            "kubernetes:yaml:parse",
```

Comment:
Parsing via invoke has gone from a previous linear runtime to O(n^3) - this could be why we're seeing increased preview times per the assumed istio helm chart usecase and https://github.com/pulumi/pulumi/issues/3665

Reply:
> Parsing via invoke has gone from a previous linear runtime to O(n^3) 

Could you explain why?  I don't think I see any change in algorithmic complexity here.

Reply:
Original, which IMO is linear:

https://github.com/pulumi/pulumi-kubernetes/blob/9d7c258073f45549b2d379c61fc470d1eb7d3c24/pkg/gen/nodejs-templates/helm/v2/helm.ts#L236-L249

Changes:

This is quadratic:
https://github.com/pulumi/pulumi-kubernetes/blob/25b5a1f880145e486868927bb543e1a145f061de/pkg/provider/provider.go#L413-L464

And above calls below which is linear:
https://github.com/pulumi/pulumi-kubernetes/blob/25b5a1f880145e486868927bb543e1a145f061de/pkg/provider/invoke_parse_yaml.go#L105-L128